gnovm/pkg/gnolang/garbage_collector.go

```diff
@@ -42,116 +68,256 @@ func (m *Machine) GarbageCollect() (left int64, ok bool) {
 	}
 
 	// Visit exceptions
-	for exception := range m.Exceptions {
-		// XXX implement for exceptions.
-		// XXX Exception is not an object,
-		// so implement a custom method and pass in vis.
+	for _, exception := range m.Exceptions {
+		stop = exception.Visit(vis, m.Store)
+		if stop {
+			return -1, false
+		}
 	}
 
 	// Return bytes remaining.
 	maxBytes, bytes := m.Alloc.Status()
 	return maxBytes - bytes, true
 }
 
-// Returns a visitor that bumps the gcCycle counter
+// Returns a visitor that bumps the GcCycle counter
 // and stops if alloc is out of memory.
-func GCVisitorFn(gcCycle int64, alloc *Alloc) Visitor {
-	vis := func(o Object) bool {
+func GCVisitorFn(gcCycle int64, alloc *Allocator) Visitor {
+	var vis func(Object) bool
+	vis = func(o Object) bool {
+		if bm.GCEnabled {
+			// The metric may either be initialized at this stage
+			// or during the previous unwrap step, depending on the flow.
+			if !bm.IsGCMeasureStarted() {
+				bm.StartGCCode(bm.VisitObject)
+			}
+		}
+
 		// Return if already measured.
 		if o.GetLastGCCycle() == gcCycle {
+			if bm.GCEnabled {
+				bm.StopGCCode()
+			}
 			return false // but don't stop
 		}
+
 		// Add object size to alloc.
 		size := o.GetShallowSize()
+		alloc.visitCount++ // count for gas calculation
+
 		alloc.Allocate(size)
 		// Stop if alloc max exceeded.
 		maxBytes, curBytes := alloc.Status()
 		if maxBytes < curBytes {
+			if bm.GCEnabled {
+				bm.StopGCCode()
+			}
 			return true
 		}
+
+		// stop metric before next visit.
+		if bm.GCEnabled {
+			bm.StopGCCode()
+		}
```

Comment:
why do we need to stop it before the next visit? 
isn't it all just visits, all continuously GC?

Reply:
removed benchmark and planned for future.